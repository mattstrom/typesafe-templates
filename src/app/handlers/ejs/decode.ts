import { NodePath } from '@babel/traverse';
import { identifier, isJSXElement, JSXElement } from '@babel/types';

import { isExprElement } from '../../tags';
import { getDataValueForAttribute, getRefValueForAttribute } from '../../helpers';
import { Handler } from '..';


export const handleDecodeElement: Handler = (path: NodePath<JSXElement>, state: any) => {
	const { children } = path.node;

	if (children.length !== 1) {
		throw new Error('$decode must contain one and only one child');
	}

	const child = path.get('children.0') as NodePath<JSXElement>;

	if (!isJSXElement(child)) {
		throw new Error('Child of $decode must be a JSX element');
	}

	const isExpr = isExprElement(child);

	const type = getDataValueForAttribute(path, 'type');
	const ref = getRefValueForAttribute(child, isExpr ? 'code' : 'value');

	const tag = `<%- decode('${type}', ${ref}) %>`;

	path.replaceWith(identifier(isExpr ? tag : `'${tag}'`));
};
