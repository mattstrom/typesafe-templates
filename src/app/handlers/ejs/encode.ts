import { NodePath } from '@babel/traverse';
import { identifier, isJSXElement, JSXElement } from '@babel/types';

import { getDataValueForAttribute, getRefValueForAttribute } from '../../helpers';
import { isExprElement } from '../../tags';
import { Handler } from '..';


export const handleEncodeElement: Handler = (path: NodePath<JSXElement>, state: any) => {
	const { children } = path.node;

	if (children.length !== 1) {
		throw new Error('$encode must contain one and only one child');
	}

	const child = path.get('children.0') as NodePath<JSXElement>;

	if (!isJSXElement(child)) {
		throw new Error('Child of $encode must be a JSX element');
	}

	const isExpr = isExprElement(child);

	const type = getDataValueForAttribute(path, 'type');
	const ref = getRefValueForAttribute(child, isExpr ? 'code' : 'value');

	const tag = `<%- encode('${type}', ${ref}) %>`;

	path.replaceWith(identifier(isExpr ? tag : `'${tag}'`));
};
