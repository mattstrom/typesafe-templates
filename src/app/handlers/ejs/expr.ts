import { NodePath } from '@babel/traverse';
import { identifier, JSXElement } from '@babel/types';
import { getRefValueForAttribute } from '../../helpers';

export function handleExprElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'code');

	path.replaceWith(
		identifier(`<%- ${ref} %>`)
	);
}
