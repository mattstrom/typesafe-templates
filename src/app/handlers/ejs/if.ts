import { NodePath } from '@babel/traverse';
import { identifier, JSXElement } from '@babel/types';

import { getRefValueForAttribute, resolveControlFlowTag } from '../../helpers';


export function handleIfElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'test');
	const { blocks, args } = resolveControlFlowTag(path, state);

	path.insertBefore(identifier(`<%_ if (${ref}) { _%>`));
	path.insertBefore(blocks);
	path.insertBefore(identifier(`<%_ } _%>`));

	path.remove();
}
