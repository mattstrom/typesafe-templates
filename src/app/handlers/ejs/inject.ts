import { NodePath } from '@babel/traverse';
import { identifier, JSXElement } from '@babel/types';
import { getRefValueForAttribute } from '../../helpers';


export function handleInjectionElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'value');

	path.replaceWith(
		identifier(`<%= ${ref} %>`)
	);
}

export function handleArrayElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'value');

	path.replaceWith(
		identifier(`<%- JSON.stringify(${ref}) %>`)
	);
}

export function handleObjectElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'value');

	path.replaceWith(
		identifier(`<%- JSON.stringify(${ref}) %>`)
	);
}

export function handleStringElement(path: NodePath<JSXElement>, state: any) {
	const ref = getRefValueForAttribute(path, 'value');

	path.replaceWith(
		identifier(`'<%- ${ref}.replace(/'/g, "\\\\\'") %>'`)
	);
}
