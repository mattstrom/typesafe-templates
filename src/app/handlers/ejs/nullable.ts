import { NodePath } from '@babel/traverse';
import {
	binaryExpression, conditionalExpression, identifier, isJSXElement,
	JSXElement, nullLiteral
} from '@babel/types';

import { getRefValueForAttribute } from '../../helpers';
import { Handler } from '..';


export const handleNullableElement: Handler = (path: NodePath<JSXElement>, state: any) => {
	const { children } = path.node;

	if (children.length !== 1) {
		throw new Error('$nullable must contain one and only one child');
	}

	const child = path.get('children.0') as NodePath<JSXElement>;

	if (!isJSXElement(child)) {
		throw new Error('Child of $nullable must be a JSX element');
	}

	const ref = getRefValueForAttribute(child, 'value');

	const node = conditionalExpression(
		binaryExpression('===', identifier(ref), identifier('undefined')),
		identifier('undefined'),
		conditionalExpression(
			binaryExpression('===', identifier(ref), nullLiteral()),
			nullLiteral(),
			child.node
		)
	);

	path.replaceWith(node);
};
