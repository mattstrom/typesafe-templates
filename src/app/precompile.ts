import * as babel from '@babel/core';
import PresetTypescript from '@babel/preset-typescript';
import PluginJsxSyntax from '@babel/plugin-syntax-jsx';
import outdent from 'outdent';

import { decode, encode } from './helpers';
import { TypesafeTemplateOptions, TypesafeTemplatePlugin } from './plugins';


export interface PrecompilerOptions {
	plugins: babel.PluginObj<any>[];
	precompile?: false | 'ejs' | 'standard';
}

const defaultOptions: PrecompilerOptions = {
	plugins: [],
	precompile: 'ejs'
};

export async function precompile(
	template: string,
	options: Partial<PrecompilerOptions> = {}
): Promise<string | null> {
	const _options = {
		...defaultOptions,
		...options
	};

	const opts: TypesafeTemplateOptions = {
		...defaultOptions,
		data: {}
	};

	const result = await babel.transformAsync(template, {
		ast: false,
		filename: 'template.tsx', // Arbitrary filename. Used by Babel to discern syntax of `contents`.
		presets: [
			PresetTypescript
		],
		plugins: [
			PluginJsxSyntax,
			[TypesafeTemplatePlugin, opts],
			..._options.plugins
		]
	});

	if (!result || !result.code) {
		return null;
	}

	return outdent`
		${result.code}
		${appendLib()}
	`;
}

function appendLib() {
	const funcs = [decode, encode];
	const lib = funcs.map(fn => `${fn}`).join('\n\n');

	const demarcation = `<%# [typesafe-templates] EJS Library Functions %>`;

	return outdent`
		${demarcation}
		<%_
			${lib}
		_%>
	`;
}
