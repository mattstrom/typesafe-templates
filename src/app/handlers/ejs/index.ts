import { TagName } from '../../tags';
import { Handler } from '..';
import {
	handleArrayElement, handleInjectionElement,
	handleObjectElement, handleStringElement
} from './inject';
import { handleDecodeElement } from './decode';
import { handleEncodeElement } from './encode';
import { handleExprElement } from './expr';
import { handleIfElement } from './if';
import { handleNullableElement } from './nullable';
import { handleRepeatElement } from './repeat';


export default new Map<TagName, Handler>([
	['$any', handleInjectionElement],
	['$array', handleArrayElement],
	['$boolean', handleInjectionElement],
	['$decode', handleDecodeElement],
	['$encode', handleEncodeElement],
	['$expr', handleExprElement],
	['$if', handleIfElement],
	['$nullable', handleNullableElement],
	['$number', handleInjectionElement],
	['$object', handleObjectElement],
	['$repeat', handleRepeatElement],
	['$string', handleStringElement]
]);
