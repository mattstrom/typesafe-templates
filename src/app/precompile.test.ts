import { render } from 'ejs';
import * as fs from 'fs';
import * as prettier from 'prettier';
import { promisify } from 'util';

import { precompile } from './precompile';


type Campus = any;

const readFile = promisify(fs.readFile);

describe('precompile()', () => {
	describe('with EJS', () => {
		let code;
		let intermediate: string;

		beforeEach(async () => {
			code = await readFile(__dirname + '/../tests/files/template.tsx', 'utf8');
			intermediate = await precompile(code) || '';
		});

		const data: Campus = {
			thing: { a: 1, b: 2 },
			buildings: [
				{
					isOpen: `(time.getHour() >= 9 && time.getHour() <= 17)`,
					rooms: [
						{ type: 'conference', number: '100', size: 200 },
						{ type: 'office', number: '201', person: 'Allison' },
						{ type: 'office', number: '202', person: 'Bob' }
					]
				},
				{
					isOpen: `(time.getHour() >= 7 && time.getHour() <= 23)`,
					rooms: [
						{ type: 'conference', number: '1', size: 75 },
						{ type: 'office', number: '2', person: 'Charlie' },
						{ type: 'office', number: '3', person: 'Denise' },
						{ type: 'closet', number: 'B1', purpose: 'utility' }
					]
				}
			]
		};

		it('should output rendered EJS', async () => {
			// Act
			const output = await render(intermediate, { $: data });
			const prettified = prettier.format(output, { parser: 'babel' });

			// Assert
			expect(prettified).toMatchSnapshot();
		});
	});

	describe('when data contain multiline strings', () => {
		it('should escape new lines as \\n', async () => {
			// Arrange
			const template = 'const str = <$string value={$.str} />';
			const data = {
				str: `
					This is a multiline
					string.
				`
			};

			const code = await precompile(template) || '';

			// Act
			const output = await render(code, { $: data });

			// Assert
			expect(output).toMatchSnapshot();
		});
	});
});
