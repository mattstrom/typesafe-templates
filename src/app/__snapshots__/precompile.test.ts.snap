// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`precompile() with EJS should output rendered EJS 1`] = `
"/// <reference path=\\"./models.d.ts\\" />
(function() {
  scheduler.plan(
    () => time.getHour() >= 9 && time.getHour() <= 17,
    schedule => {
      schedule.reserve(\\"100\\", 200);
      schedule.assign(\\"201\\", \\"Allison\\");
      schedule.assign(\\"202\\", \\"Bob\\");
    }
  );
  scheduler.plan(
    () => time.getHour() >= 7 && time.getHour() <= 23,
    schedule => {
      schedule.reserve(\\"1\\", 75);
      schedule.assign(\\"2\\", \\"Charlie\\");
      schedule.assign(\\"3\\", \\"Denise\\");
      schedule.include();
    }
  );
})();
"
`;
